// @storybook/ui/dist/modules/ui/components/layout/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _global = require('global');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactSplitPane = require('react-split-pane');

var _reactSplitPane2 = _interopRequireDefault(_reactSplitPane);

var _usplit = require('./usplit');

var _usplit2 = _interopRequireDefault(_usplit);

var _dimensions = require('./dimensions');

var _dimensions2 = _interopRequireDefault(_dimensions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var rootStyle = {
  height: '100vh',
  backgroundColor: '#F7F7F7'
};

var leftPanelStyle = function leftPanelStyle(showLeftPanel, leftPanelOnTop) {
  return {
    width: '100%',
    display: showLeftPanel ? 'flex' : 'none',
    flexDirection: leftPanelOnTop ? 'column' : 'row',
    alignItems: 'stretch',
    paddingRight: leftPanelOnTop ? 10 : 0
  };
};

var downPanelStyle = function downPanelStyle(showDownPanel, downPanelInRight) {
  return {
    display: showDownPanel ? 'flex' : 'none',
    flexDirection: downPanelInRight ? 'row' : 'column',
    alignItems: 'stretch',
    position: 'absolute',
    width: '100%',
    height: '100%',
    padding: downPanelInRight ? '5px 10px 10px 0' : '0px 10px 10px 0',
    boxSizing: 'border-box'
  };
};

var resizerCursor = function resizerCursor(isVert) {
  return isVert ? 'col-resize' : 'row-resize';
};

var storiesResizerStyle = function storiesResizerStyle(showLeftPanel, leftPanelOnTop) {
  return {
    cursor: showLeftPanel ? resizerCursor(!leftPanelOnTop) : undefined,
    height: leftPanelOnTop ? 10 : 'auto',
    width: leftPanelOnTop ? '100%' : 10,
    zIndex: 1
  };
};

var addonResizerStyle = function addonResizerStyle(showDownPanel, downPanelInRight) {
  return {
    cursor: showDownPanel ? resizerCursor(downPanelInRight) : undefined,
    height: downPanelInRight ? '100%' : 10,
    width: downPanelInRight ? 10 : '100%',
    zIndex: 1
  };
};

var contentPanelStyle = function contentPanelStyle(downPanelInRight, leftPanelOnTop) {
  return {
    position: 'absolute',
    boxSizing: 'border-box',
    width: '100%',
    height: '100%',
    padding: downPanelInRight ? '10px 2px 10px 0' : '10px 10px 2px 0',
    paddingTop: leftPanelOnTop ? 0 : 10
  };
};

var normalPreviewStyle = {
  width: '100%',
  height: '100%',
  backgroundColor: '#FFF',
  border: '1px solid #ECECEC',
  borderRadius: 4
};

var fullScreenPreviewStyle = {
  position: 'fixed',
  left: '0px',
  right: '0px',
  top: '0px',
  zIndex: 1,
  backgroundColor: '#FFF',
  height: '100%',
  width: '100%',
  border: 0,
  margin: 0,
  padding: 0,
  WebkitOverflowScrolling: 'touch'
};

var defaultSizes = {
  addonPanel: {
    down: 200,
    right: 400
  },
  storiesPanel: {
    left: 250,
    top: 400
  }
};

var onDragStart = function onDragStart() {
  return _global.document.body.classList.add('dragging');
};

var onDragEnd = function onDragEnd() {
  return _global.document.body.classList.remove('dragging');
};

var saveSizes = function saveSizes(sizes) {
  try {
    _global.localStorage.setItem('panelSizes', (0, _stringify2.default)(sizes));
    return true;
  } catch (e) {
    return false;
  }
};

var getSavedSizes = function getSavedSizes(sizes) {
  try {
    var panelSizes = _global.localStorage.getItem('panelSizes');
    if (panelSizes) {
      return JSON.parse(panelSizes);
    }
    saveSizes(sizes);
    return sizes;
  } catch (e) {
    saveSizes(sizes);
    return sizes;
  }
};

var Layout = function (_React$Component) {
  (0, _inherits3.default)(Layout, _React$Component);

  function Layout(props) {
    (0, _classCallCheck3.default)(this, Layout);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Layout.__proto__ || (0, _getPrototypeOf2.default)(Layout)).call(this, props));

    _this.layerSizes = getSavedSizes(defaultSizes);

    _this.state = {
      previewPanelDimensions: {
        height: 0,
        width: 0
      }
    };

    _this.onResize = _this.onResize.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(Layout, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      _global.window.addEventListener('resize', this.onResize);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      _global.window.removeEventListener('resize', this.onResize);
    }
  }, {
    key: 'onResize',
    value: function onResize(pane, mode) {
      var _this2 = this;

      return function (size) {
        _this2.layerSizes[pane][mode] = size;
        saveSizes(_this2.layerSizes);

        var _previewPanelRef = _this2.previewPanelRef,
            clientWidth = _previewPanelRef.clientWidth,
            clientHeight = _previewPanelRef.clientHeight;


        _this2.setState({
          previewPanelDimensions: {
            width: clientWidth,
            height: clientHeight
          }
        });
      };
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          goFullScreen = _props.goFullScreen,
          showLeftPanel = _props.showLeftPanel,
          showDownPanel = _props.showDownPanel,
          downPanelInRight = _props.downPanelInRight,
          downPanel = _props.downPanel,
          leftPanel = _props.leftPanel,
          preview = _props.preview;
      var previewPanelDimensions = this.state.previewPanelDimensions;


      var leftPanelOnTop = false;

      var previewStyle = normalPreviewStyle;

      if (goFullScreen) {
        previewStyle = fullScreenPreviewStyle;
      }

      var sizes = getSavedSizes(this.layerSizes);

      var leftPanelDefaultSize = !leftPanelOnTop ? sizes.storiesPanel.left : sizes.storiesPanel.top;
      var downPanelDefaultSize = !downPanelInRight ? sizes.addonPanel.down : sizes.addonPanel.right;

      var addonSplit = downPanelInRight ? 'vertical' : 'horizontal';
      var storiesSplit = leftPanelOnTop ? 'horizontal' : 'vertical';

      return _react2.default.createElement(
        'div',
        { style: rootStyle },
        _react2.default.createElement(
          _reactSplitPane2.default,
          {
            split: storiesSplit,
            allowResize: showLeftPanel,
            minSize: 150,
            maxSize: -400,
            size: showLeftPanel ? leftPanelDefaultSize : 1,
            defaultSize: leftPanelDefaultSize,
            resizerStyle: storiesResizerStyle(showLeftPanel, leftPanelOnTop),
            onDragStarted: onDragStart,
            onDragFinished: onDragEnd,
            onChange: this.onResize('storiesPanel', leftPanelOnTop ? 'top' : 'left')
          },
          _react2.default.createElement(
            'div',
            { style: leftPanelStyle(showLeftPanel, leftPanelOnTop) },
            _react2.default.createElement(
              'div',
              { style: { flexGrow: 1, height: '100%', width: '100%' } },
              leftPanel()
            ),
            _react2.default.createElement(_usplit2.default, { shift: 5, split: storiesSplit })
          ),
          _react2.default.createElement(
            _reactSplitPane2.default,
            {
              split: addonSplit,
              allowResize: showDownPanel,
              primary: 'second',
              minSize: downPanelInRight ? 200 : 100,
              maxSize: -200,
              size: showDownPanel ? downPanelDefaultSize : 1,
              defaultSize: downPanelDefaultSize,
              resizerStyle: addonResizerStyle(showDownPanel, downPanelInRight),
              onDragStarted: onDragStart,
              onDragFinished: onDragEnd,
              onChange: this.onResize('addonPanel', downPanelInRight ? 'right' : 'down')
            },
            _react2.default.createElement(
              'div',
              { style: contentPanelStyle(downPanelInRight, leftPanelOnTop) },
              _react2.default.createElement(
                'div',
                {
                  style: previewStyle,
                  ref: function ref(_ref) {
                    _this3.previewPanelRef = _ref;
                  }
                },
                preview()
              ),
              _react2.default.createElement(_dimensions2.default, previewPanelDimensions)
            ),
            _react2.default.createElement(
              'div',
              { style: downPanelStyle(showDownPanel, downPanelInRight) },
              _react2.default.createElement(_usplit2.default, { shift: -5, split: addonSplit }),
              downPanel()
            )
          )
        )
      );
    }
  }]);
  return Layout;
}(_react2.default.Component);

Layout.propTypes = {
  showLeftPanel: _propTypes2.default.bool.isRequired,
  showDownPanel: _propTypes2.default.bool.isRequired,
  goFullScreen: _propTypes2.default.bool.isRequired,
  leftPanel: _propTypes2.default.func.isRequired,
  preview: _propTypes2.default.func.isRequired,
  downPanel: _propTypes2.default.func.isRequired,
  downPanelInRight: _propTypes2.default.bool.isRequired
};

exports.default = Layout;
